src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -2733,11 +2733,13 @@ private bool Fix(int iParam, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo
             if (containsFunctionTypes(lower) &&
                 (containsNonFunctionTypes(lower) || containsNonFunctionTypes(exact) || containsNonFunctionTypes(upper)))
             {
-                lower = removeTypes(lower, static type => isFunctionType(type, out _));
+                removeTypes(lower, static type => isFunctionType(type, out _));
```

Comment:
>removeTypes(lower, static type => isFunctionType(type, out _)); [](http://example.com/codeflow?start=16&length=63)

The fact that the function now "destroys" the input makes me feel uncomfortable. Could we instead pass the filter delegate along with the ```lower``` and apply it when we enumerate its items, rather than removing the items from the set?

Reply:
@AlekseyTs In the original code, `lower` would have been replaced with the new HashSet returned from `removeTypes`, so it's kinda destroyed as well. Yeah I understand that the original instance isn't modified/destroyed, but I mean it's thrown away and a new HashSet is assigned.

I don't see how this can break anything.

Reply:
> In the original code, lower would have been replaced with the new HashSet returned from removeTypes, so it's kinda destroyed as well. 

Not really, original code didn't mess with the caller's hash set. I.e. didn't change its content,

Reply:
I see your concern now. Applying the predicate while enumerating could even be more efficient anyway.

Reply:
This possibly risks evaluating the predicate twice, but I think that is still better than the aggressive hashset allocations.